src/mlpack/tests/CMakeLists.txt

```diff
@@ -272,6 +272,6 @@ add_custom_command(TARGET mlpack_test
   WORKING_DIRECTORY ${PROJECT_BINARY_DIR}
 )
 
-add_test(NAME "catch_test" COMMAND mlpack_test WORKING_DIRECTORY ${CMAKE_BINARY_DIR})
+add_test(NAME "catch_test" COMMAND mlpack_test WORKING_DIRECTORY ${CMAKE_CURRENT_BINARY_DIR})
```

Comment:
I didn't notice this until the very end, but throughout this PR you replaced `CMAKE_SOURCE_DIR` with `CMAKE_CURRENT_SOURCE_DIR`, and `CMAKE_BINARY_DIR` with `CMAKE_CURRENT_BINARY_DIR`.  But the diff here showed a variable `PROJECT_BINARY_DIR`... see https://cmake.org/cmake/help/latest/variable/PROJECT_BINARY_DIR.html

There is also a variable `PROJECT_SOURCE_DIR`: https://cmake.org/cmake/help/latest/variable/PROJECT_SOURCE_DIR.html

I suspect that using those two variables instead, everything would "just work" as those paths would be the roots of the mlpack project directory.  Do you want to try that?